### ISSUE TITLE
[session-watcher] Fail to start on remote

### ISSUE THREAD

**Issue by pat-s** [REditorSupport/vscode-R#259]
Config:

- Startup with custom R interpreter and project based `.Rprofile`
- `r.sessionwatcher` enabled

```r
# vscode
Sys.setenv(TERM_PROGRAM = "vscode")
source(file.path(Sys.getenv(if (.Platform$OS.type == "windows") "USERPROFILE" else "HOME"), ".vscode-R", "init.R"))
```

set in project root's `.Rprofile`

I even added a `cat()` statement in `~/.vscode-R/init.R` to track whether the code is being executed.

```
radian                                                                                                      
vscode session watcher: success
R version 3.6.1 (2019-07-05) -- "Action of the Toes"
Platform: x86_64-pc-linux-gnu (64-bit)
```

However, the R session is still not attached in vscode.
The same happens outside of the project (i.e. without a custom .Rprofile).

Running "R: Create R Terminal" errors - but so quickly that I cannot see the reason. I also cannot find a log for this.

# More info

- `R_HOME = /opt/spack/opt/spack/linux-centos7-x86_64/gcc-9.2.0/r-3.6.1-j25wr6zcofibs2zfjwg37357rjj26lqb/rlib/R`
- radian v0.4.4

**Comment by renkun-ken** [comment]
Thanks for your reporting!

Would you like to run the following command in your R session and post the output:

```r
ls.str(getOption("vscodeR"))
```

Make sure your workspace root is the working directory where you start your R session, or otherwise the R session watcher would not know the session to attach.

**Comment by pat-s** [comment]
radian started from $HOME. vscode status: "R: (not attached)".

```r
ls.str(getOption("vscodeR"))    
                                                                                 
attach : function ()  
dataview : function (x, title)  
dataview_data_type : function (x)  
dataview_table : function (data)  
dir :  chr "/home/patrick/.vscode/vscode-R"
dir_plot_history :  chr "/tmp/Rtmp5fZj9Q/images"
dir_session :  chr "/home/patrick/.vscode/vscode-R/93795"
globalenv_file :  chr "/home/patrick/.vscode/vscode-R/93795/globalenv.json"
new_plot : function ()  
pid :  int 93795
platform : List of 8
 $ OS.type   : chr "unix"
 $ file.sep  : chr "/"
 $ dynlib.ext: chr ".so"
 $ GUI       : chr "vscode"
 $ endian    : chr "little"
 $ pkgType   : chr "source"
 $ path.sep  : chr ":"
 $ r_arch    : chr ""
plot_file :  chr "/home/patrick/.vscode/vscode-R/93795/plot.png"
plot_history_file :  NULL
plot_updated :  logi FALSE
rebind : function (sym, value, ns)  
respond : function (command, ...)  
response_file :  chr "/home/patrick/.vscode/vscode-R/response.log"
tempdir :  chr "/tmp/Rtmp5fZj9Q"
unbox : function (x)  
update : function (...)  
```

**Comment by renkun-ken** [comment]
OK. Then is your workspace root in VSCode also your home directory?

**Comment by pat-s** [comment]
Ah! Not in the example above. Doing so in R started from the workspace root results in 


```r
ls.str(getOption("vscodeR"))                                                                                     
Error in as.environment(pos) : using 'as.environment(NULL)' is defunct
```

(R 3.6.1)

**Comment by renkun-ken** [comment]
You said there's a `.Rprofile` in your workspace root and it contains

```r
Sys.setenv(TERM_PROGRAM = "vscode")
source(file.path(Sys.getenv(if (.Platform$OS.type == "windows") "USERPROFILE" else "HOME"), ".vscode-R", "init.R"))
```

and `using 'as.environment(NULL)' is defunct` suggests that the project `.Rprofile` is not sourced at all.

**Comment by pat-s** [comment]
It's sourced because I can see the {renv} header after startup. I am calling the vscode lines after `source("renv/activate.R")` in the workspace root `.Rprofile` - could this mess things up?

Also

```r
Sys.getenv("TERM_PROGRAM")    
[1] "vscode"
```

but yeah - the custom cat() output I added to `init.R` is missing here - hence it is not being executed (correctly).

Calling `source(file.path(Sys.getenv(if (.Platform$OS.type == "windows") "USERPROFILE" else "HOME"), ".vscode-R", "init.R"))` after startup attaches the session.

-> Turns out I have to call the vscode lines before `source("renv/activate.R")`

Thanks for helping - turns out its not related to an issue on your side but looks like renv is getting into the way somehow.

**Comment by renkun-ken** [comment]
If your `renv/activate.R` script contains code that changes working directory, then the session watcher won't work.

**Comment by pat-s** [comment]
> If your renv/activate.R script contains code that changes working directory, then the session watcher won't work.

It does [a lot](https://github.com/rstudio/renv/blob/5f820844c3512e223b8272779ce0ecc308fce0e6/renv/activate.R) - haven't looked into it in detail.

Maybe this piece of information could be added to the wiki (I can't, it is not publicly editable) - {renv} is a common thing to do for project management and it will probably come up again.

**Comment by renkun-ken** [comment]
Thanks for pointing to renv. I'll take a closer look at it and I'll update the wiki on this soon.

**Comment by markk-fmi** [comment]
I have a similar issue in that session-watcher works fine locally, but on remote all the R-related settings.json parameters are greyed out and have mouse-over notice of "Unknown Configuration Setting".  Here are the greyed out parameters from `settings.json`:
```
    "r.sessionWather": true,
    "r.rterm.linux": "/home/mkennedy/.venv/analysis/bin/radian",
    "r.rterm.mac": "/Users/mkennedy/opt/miniconda3/bin/radian",
    "r.rterm.option": [
        "--no-save",
        "--no-restore",
        "--r-binary=/usr/local/bin/R"
    ],
    "r.bracketedPaste": true,
    "r.sessionWatcher": true,
    "r.lsp.debug": true,
    "r.lsp.diagnostics": true,
```

And I'm unable to do cmd-return (mac) to connect to radian?

Might you have any suggestions about how to configure vscode for remote R development with the session watcher?

**Comment by renkun-ken** [comment]
>but on remote all the R-related settings.json parameters are greyed out and have mouse-over notice of "Unknown Configuration Setting"

Looks like you have not installed the extensions on your remote machine. Please know that local and remote machine have separate spaces of installed extensions. You need to go to the Extensions pane and choose to install a subset of the extensions installed locally to the remote machine.

**Comment by markk-fmi** [comment]
Thanks for that insight!  I didn't understand that extensions installed were not available on remotes.  So I followed your advice and there's good news and some challenges still.

I had gotten my local version to work by including `"--r-binary=/usr/local/bin/R"` in the `settings.json` `"r.rterm.option"` list.  This, of course, did not work on the remote as there is a different path.  There doesn't seem to be a means to specify options by os as there is for `rterm`, right?  So I deleted this from the `r.rterm.option` list.  The good news is the watcher works on the remote now after your suggestion of enabling the extensions on the remote and deleting the incorrect path option.  Interestingly no path to R was needed.

The not so good news is that it does not work locally without the R-path option in r.rterm.option.  I'm actually confused about why this is needed on the local mac laptop.  The gripe when I try to do a cmd-return in vscode is: 
```
"/Users/myname/opt/miniconda3/bin/radian '--no-save', '--no-restore'" failed to launch (exit code: 1).
```

But if I simply run this on the command line it works fine and gives me the radian prompt.  So I'm not sure what is different when radian is executed in vscode?

I note that prior to the exit code error message the terminal flashes another error, which it clears, unfortunately.  I was able to take a video of it and found that it is a python traceback that goes:
```
radian, line 8, in <module>
radian/__init__.py, line 44, in main
python3.8/site-packages/rchitect/utiles.py, line 84, in Rhome
   raise RuntimeError("Cannot determine R HOME")
```

In the `Rhome` method the exception is thrown if this call `rhome = getRhome("R")` returns None.  But the `getRhome` method uses this line to get the value:
```
rhome = subprocess.check_output([path, "RHOME"]).decode("utf-8").strip()
```

But I can run this line in my python terminal and it works fine:
```
>>> subprocess.check_output(["R", "RHOME"]).decode("utf-8").strip()
'/Library/Frameworks/R.framework/Resources'
>>> 
```

So might you have any thoughts about why in vscode this exception is being raised?
